### ISSUE TITLE
no response 

### ISSUE THREAD

**Issue by mrasmith** [N1ghteyes/cpanel-UAPI-php-class#5]
Hi , thanks for sharing this. When I try this class, I get no response and no errors .. have I missed something obvious ? I've included what I'm using below..

Thanks for any help. 
 
```
// include class
include "includes/cpanel-UAPI-php-class/cpaneluapi.class.php"; //include the class file

// credentials
include_once("includes/cpanel_inc.php");;

// connect cpanel api
$capi = new cpanelAPI( $account, $passwd, $ip ); // instantiate the object

$response = $capi->uapi->Mysql->get_restrictions(); 
print_r($response);
```

**Comment by N1ghteyes** [comment]
Hi mrasmith,

Do you get anything if you do this?
```
$response = $capi->uapi->Mysql->get_restrictions(); 
print_r($capi->getError());
```
There isn't a whole lot of error handling in the class, its on the todo list the i get a chance.

**Comment by mrasmith** [comment]
Thanks for the response, no unfortunately that gives no output either.

**Comment by N1ghteyes** [comment]
Hmm ok, next thought then - is the IP for cPanel or WHM? and is it HTTPS or just HTTP?

**Comment by mrasmith** [comment]
cPanel and WHM, it’s a dedicated server, just HTTP

**Comment by N1ghteyes** [comment]
Ah ok, so its might be that the class assumes HTTPS, which means that the request will use the wrong port and protocol.

try this:

```
// include class
include "includes/cpanel-UAPI-php-class/cpaneluapi.class.php"; //include the class file

// credentials
include_once("includes/cpanel_inc.php");;

// connect cpanel api
$capi = new cpanelAPI( $account, $passwd, $ip ); // instantiate the object

//no SSL
$capi->ssl = FALSE;

//get response
$response = $capi->uapi->Mysql->get_restrictions(); 
print_r($response);
```

**Comment by N1ghteyes** [comment]
Oh you'll need to change the port manually as well it seems. I should fix that. Updated code below:

```
// include class
include "includes/cpanel-UAPI-php-class/cpaneluapi.class.php"; //include the class file

// credentials
include_once("includes/cpanel_inc.php");;

// connect cpanel api
$capi = new cpanelAPI( $account, $passwd, $ip ); // instantiate the object

//no SSL
$capi->ssl = FALSE;
$capi->port = 2082;  //Check this!

//get response
$response = $capi->uapi->Mysql->get_restrictions(); 
print_r($response);
```

**Comment by mrasmith** [comment]
added both of those and still no output ! I appreciate your help with this.

**Comment by N1ghteyes** [comment]
could you paste the response of this for me?

```
// include class
include "includes/cpanel-UAPI-php-class/cpaneluapi.class.php"; //include the class file

// credentials
include_once("includes/cpanel_inc.php");;

// connect cpanel api
$capi = new cpanelAPI( $account, $passwd, $ip ); // instantiate the object

//no SSL
$capi->ssl = FALSE;
$capi->port = 2082;  //Check this!

//get response
$response = $capi->uapi->Mysql->get_restrictions(); 
print_r($capi->getLastRequest());
```

**Comment by mrasmith** [comment]
I've x'd out the ip address 

`http://xxx.xxx.xxx.xxx:2082/execute/Mysql/get_restrictions?`

**Comment by N1ghteyes** [comment]
that looks correct. can you confirm that the port you access cPanel on is 2082? if you're not sure just try navigating to http://xxx.xxx.xxx.xxx:2082 directly.

If thats working as expected i'll find some time to setup some test scripts, see if i can reproduce the problem myself :).

**Comment by mrasmith** [comment]
yes that brings me to my cpanel login

**Comment by mrasmith** [comment]
Thanks a lot for your help

**Comment by N1ghteyes** [comment]
In that case i'll have to run some proper tests :).

If you don't get a reply from me by this time tomorrow, feel free to prod!

**Comment by mrasmith** [comment]
would it matter that I have 2FA enabled on my account ? I assumed it wouldn't matter because I use another class for the API1 & 2 without any issues.

**Comment by N1ghteyes** [comment]
oh, it might. API 1 and 2 often include a cPanel api file on the server and communicate through that. UAPI makes requests over HTTP and thus must pass authentication details with every request. 2FA may well interfere with that process.

There should be a workaround, i'll dig in the docs.

**Comment by mrasmith** [comment]
Great, thanks again.

**Comment by N1ghteyes** [comment]
Ok, so i found the docs - i'll need to make some amends to the class as 2FA requires an additional parameter and some extra data. Doesn't look like too much of a pain so i'll get to it now.

**Comment by N1ghteyes** [comment]
ok so, I've updated the class. Its a bit of a blind guess functionality wise, but please give that a go. You'll need to pass the 2FA secret into the class when instantiating it. E.g.

```
// connect cpanel api
$capi = new cpanelAPI( $account, $passwd, $ip, $secret); // instantiate the object
```

let me know if it works, or if not, what breaks for you. I haven't had time to test

**Comment by mrasmith** [comment]
Thanks a million, I'm not in a position to test right now but will let you know as soon as I do.

**Comment by mrasmith** [comment]
Thanks N1ghteyes for this but unfortunately still no output .

**Comment by N1ghteyes** [comment]
Hi Mrasmith, could you disable 2FA and test? just to be sure that is the issue.

Thanks

**Comment by mrasmith** [comment]
ok done that and it now works, so I guess we know where the problem is !

**Comment by N1ghteyes** [comment]
In that case i'll have to setup 2FA on an account somewhere and play with it :). I'll update here when i get somewhere - might not be for a few days though.

**Comment by mrasmith** [comment]
no problem, thanks for your help

**Comment by mrasmith** [comment]
You did mention I should feel free to prod ,, so here's a gentle prod ! Thanks.

**Comment by N1ghteyes** [comment]
@mrasmith Yep - sorry been a busy few weeks! I've been looking into the problem, there appears to be a gap in the docs for this. I'll keep digging though, i will also send a note to cPanel support and see if they can shed any light on it.

**Comment by N1ghteyes** [comment]
Just a quick update to say there is no update. Waiting on a response from cPanel!

**Comment by ViperJim** [comment]
I am having the same problem but I am using https not using 2Fa.
I get no output.
I tried the test from above and did get the following output, "http://xxxxxxxxxxxxx:2082/execute/Mysql/get_restrictions?"
When logging in manually I use, https://xxxxxxxxxxxx:2083.
I am using,  $capi = new cpanelAPI( $account, $passwd, $ip, $secret); // instantiate the object
Still no output.

**Comment by N1ghteyes** [comment]
@joxley could you open a separate issue and post the exact block of code that's getting no output? (Without any access info of course).

The lack of output in this instance was specifically 2FA related.

**Comment by grahams2020** [comment]
Hi there, have you had any luck with this? We've been using your library for a while now on our web app (which is great btw) and it's worked great, up until we updated our security and installed 2FA on cPanel. Now the uapi calls don't work, but api2 calls do. Could redo code to use all api2 requests, but would rather use the latest api from cPanel.

**Comment by N1ghteyes** [comment]
@grahams2020 Whoops, sorry missed the post.

I'll chase, its been sitting on the back burner for a while and i haven't heard back from cPanel. In theory, whats in place follows their documentation but the authentication fails. I suspect a bug in the UAPI or perhaps old / incorrect docs.

Even the email address to add new third party libraries to their list doesn't work!

**Comment by marasi** [comment]
Hi. I'm having no return from script at all.

**Comment by N1ghteyes** [comment]
@marasi Are you also using 2FA? if so, i still can't get anything out of cPanel so the issue is currently parked until i have time to chase them up on it.

If you're not, then please could you open a new issue, include your code so we can debug.

Thanks

**Comment by N1ghteyes** [comment]
@marasi @mrasmith 

Hi Both, A recent amend was pushed which might finally solve this issue. If you have a chance to check that would be awesome. If in the mean time you've found a fix please let me know!

Thanks

**Comment by N1ghteyes** [comment]
Closing since this hasn't come up again in a while.
